sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -0,0 +1,316 @@
+          // We only change (scale down) the number of distinct values if the number of rows
+          // decreases after join, because join won't produce new values even if the number of
+          // rows increases.
+          val newNdv = if (join.left.outputSet.contains(a) && leftRatio < 1) {
+            ceil(BigDecimal(oldNdv) * leftRatio)
+          } else if (join.right.outputSet.contains(a) && rightRatio < 1) {
+            ceil(BigDecimal(oldNdv) * rightRatio)
+          } else {
+            oldNdv
+          }
+          // TODO: support nullCount updates for specific outer joins
+          outputAttrStats.put(a, oldCS.copy(distinctCount = newNdv))
+        }
+      }
+    }
+    AttributeMap(outputAttrStats.toSeq)
+  }
+
+  /** Update intersected column stats for join keys. */
+  private def updateIntersectedStats(
+      joinKeyPairs: Seq[(AttributeReference, AttributeReference)],
+      leftStats: Statistics,
+      rightStats: Statistics): AttributeMap[ColumnStat] = {
+    val intersectedStats = new mutable.HashMap[Attribute, ColumnStat]()
+    joinKeyPairs.foreach { case (leftKey, rightKey) =>
+      // Do estimation if we have enough statistics
+      if (columnStatsExist((leftStats, leftKey), (rightStats, rightKey))) {
+        // Check if the two sides are disjoint
+        val leftKeyStats = leftStats.attributeStats(leftKey)
+        val rightKeyStats = rightStats.attributeStats(rightKey)
+        val lRange = Range(leftKeyStats.min, leftKeyStats.max, leftKey.dataType)
+        val rRange = Range(rightKeyStats.min, rightKeyStats.max, rightKey.dataType)
+        if (Range.isIntersected(lRange, rRange)) {
+          // Update intersected column stats
+          val minNdv = leftKeyStats.distinctCount.min(rightKeyStats.distinctCount)
+          val (newMin1, newMax1, newMin2, newMax2) =
+            Range.intersect(lRange, rRange, leftKey.dataType, rightKey.dataType)
+          intersectedStats.put(leftKey, intersectedColumnStat(leftKeyStats, minNdv,
+            newMin1, newMax1))
+          intersectedStats.put(rightKey, intersectedColumnStat(rightKeyStats, minNdv,
+            newMin2, newMax2))
+        }
+      }
+    }
+    AttributeMap(intersectedStats.toSeq)
+  }
+
+  private def emptyColumnStat(dataType: DataType): ColumnStat = {
+    ColumnStat(distinctCount = 0, min = None, max = None, nullCount = 0,
+      avgLen = dataType.defaultSize, maxLen = dataType.defaultSize)
+  }
+
+  private def intersectedColumnStat(
+      origin: ColumnStat,
+      newDistinctCount: BigInt,
+      newMin: Option[Any],
+      newMax: Option[Any]): ColumnStat = {
+    origin.copy(distinctCount = newDistinctCount, min = newMin, max = newMax, nullCount = 0)
+  }
+
+  private def extractJoinKeys(
+      leftKeys: Seq[Expression],
+      rightKeys: Seq[Expression]): Seq[(AttributeReference, AttributeReference)] = {
+    leftKeys.zip(rightKeys).flatMap {
+      case (lk: AttributeReference, rk: AttributeReference) => Some((lk, rk))
```

Comment:
we can check column stats existence here, so that we don't need to do `columnStatsExist((leftStats, leftKey), (rightStats, rightKey))` again and again later.